returnn/tf/layers/signal_processing.py

```diff
@@ -607,6 +607,72 @@ def get_out_data_from_opts(cls, fft_size, use_rfft=True, nr_of_channels=1, **kwa
             .copy_template(dtype="complex64"))
 
 
+class StftLayer(_ConcatInputLayer):
+  """
+  A generic STFT layer.
+  """
+
+  layer_class = "stft"
+  recurrent = True  # we should not shuffle in the time-dimension
+
+  # noinspection PyUnusedLocal
+  def __init__(self, frame_shift, frame_size, fft_size=None, **kwargs):
+    """
+    :param int frame_shift: frame shift for STFT
+    :param int frame_size: frame size for STFT
+    :param Optional[int] fft_size: size of the FFT to apply
+    """
+    fft_size = fft_size or frame_size
+    assert "n_out" not in kwargs
+    assert "out_type" not in kwargs
+    super(StftLayer, self).__init__(**kwargs)
+    assert not self.input_data.sparse
+    assert self.input_data.have_batch_axis()
+    assert self.input_data.have_time_axis()
+    out_dim = fft_size // 2 + 1
+    in_dim = self.input_data.feature_dim_or_sparse_dim
+    input_data, num_batch_dims = ConvLayer.transform_input(
+      self.input_data, network=self.network, in_dim=in_dim)
+    # We want to prepare the input data such that the batch-dim(s) is the very first,
+    # the feature-dim is right after batch-dim ("NCHW"), and the spatial dim is last.
+    if self.output.feature_dim_axis == num_batch_dims:
+      input_data = input_data.copy_with_feature_dim_axis(num_batch_dims)
+    else:
+      input_data = input_data.copy_with_feature_dim_axis(-1)
+    x = input_data.placeholder
+    # squeeze feature axis
+    assert input_data.batch_shape[input_data.feature_dim_axis] == 1, "only implemented for single channel"
+    x = tf.squeeze(x, axis=input_data.feature_dim_axis)
+    extended_batch_shape = None
+    if num_batch_dims > 1:
+      x_shape = tf.shape(x)
+      extended_batch_shape = x_shape[:num_batch_dims]
+      x = tf.reshape(x, tf.concat([[-1], x_shape[num_batch_dims:]], axis=0))  # merge all batch dims
+    y = tf.signal.stft(x, frame_length=frame_size, frame_step=frame_shift, fft_length=fft_size)
+    # y shape is [batch] + spatial_dims + [n_out].
+    if num_batch_dims > 1:
+      y = tf.reshape(y, tf.concat([extended_batch_shape, tf.shape(y)[1:]], axis=0))
+    self.output.placeholder = y
+
+  @classmethod
+  def get_out_data_from_opts(cls, name, sources, network, frame_shift, frame_size, fft_size=None, **kwargs):
+    """
+    :param str name:
+    :param list[LayerBase] sources:
+    :param returnn.tf.network.TFNetwork network:
+    :param int frame_shift: frame shift for STFT
+    :param int frame_size: frame size for STFT
+    :param Optional[int] fft_size: size of the FFT to apply
+    :rtype: Data
+    """
+    fft_size = fft_size or frame_size
+    data = ConvLayer.get_out_data_from_opts(
+      name=name, sources=sources, network=network, filter_size=[frame_size], strides=[frame_shift],
+      padding="valid", n_out=fft_size // 2 + 1)
```

Comment:
According to [the docs](https://www.tensorflow.org/api_docs/python/tf/signal/stft), the output data would always have feature-last, right?

Reply:
Yes that's right. Does that change anything here?